### PR TITLE
feat: implement ICS-721 NFT transfer application

### LINKED ISSUE #346: Implement ICS 721: NFT transfer on `ibc-rs`
## Summary

Implement iCS 721: NFT transfer on `ibc-rs`

## Problem Definition

At Composable, we are currently finishing up our Centauri Bridge, which will connect the Polkadot with Cosmos. It is important
for us to be able to support NFT transfers to give those chains that are using this feature through IBC.

## Proposal

Ideally follow the [spec](https://github.com/cosmos/ibc/blob/main/spec/app/ics-721-nft-transfer/README.md) and the [reference implementation written in go](https://github.com/bianjieai/nft-transfer), and a [reference implementation in CosmWasm](https://github.com/public-awesome/ics721)

### Happy path
NFT being transferred from chain A to B, and then back from B to A.

### Less happy path
NFT transfer timing out, and not being transferred to chain B from A.
NFT transfer succeeding from A to B, but timing out from B to A.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #346

## Description

This PR introduces the implementation of the ICS-721 NFT transfer application on the current `feat/ics721-impl` feature branch to the `main`. It encompasses the following PRs:

- https://github.com/cosmos/ibc-rs/pull/1012
- https://github.com/cosmos/ibc-rs/pull/1020
- https://github.com/cosmos/ibc-rs/pull/1027
- https://github.com/cosmos/ibc-rs/pull/1032
- https://github.com/cosmos/ibc-rs/pull/1038
- https://github.com/cosmos/ibc-rs/pull/1039

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-apps/ics20-transfer/types/src/msgs/transfer.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//! Defines the token transfer message type
+//! Defines the Non-Fungible Token Transfer message type
```
Comment (rnbguy) [inline ibc-apps/ics20-transfer/types/src/msgs/transfer.rs]: is this a mistake? this is ics20 source code.

File: ibc-apps/ics20-transfer/types/src/msgs/transfer.rs [hunk 2]
```diff
@@ -15,7 +15,7 @@ use crate::packet::PacketData;
 
 pub(crate) const TYPE_URL: &str = "/ibc.applications.transfer.v1.MsgTransfer";
 
-/// Message used to build an ICS20 token transfer packet.
+/// Message used to build an ICS-721 Non-Fungible Token Transfer packet.
```
Comment (rnbguy) [inline ibc-apps/ics20-transfer/types/src/msgs/transfer.rs]: mention of ics721 inside ics20 source code.

File: ibc-apps/ics721-nft-transfer/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    "std",
+]
+borsh = [
+    "ibc-app-nft-transfer-types/borsh",
+    "ibc-core/borsh",
+]
+parity-scale-codec = [
+    "ibc-app-nft-transfer-types/parity-scale-codec",
+    "ibc-core/parity-scale-codec",
+]
```
Comment (rnbguy) [inline ibc-apps/ics721-nft-transfer/Cargo.toml]: Add line ending.

```suggestion
]

```

**Review by rnbguy — APPROVED** [review]
LGTM :+1:

I approved the PR but I think the changes in ics20 source code are mistakes.

File: ibc-apps/ics20-transfer/types/src/msgs/transfer.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics20-transfer/types/src/msgs/transfer.rs]: Oops! [9333f77](https://github.com/cosmos/ibc-rs/pull/1053/commits/9333f775d3670c1f1922eb9b110015db4b7d5330)

File: ibc-apps/ics20-transfer/types/src/msgs/transfer.rs (hunk 2 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics20-transfer/types/src/msgs/transfer.rs]: [9333f77](https://github.com/cosmos/ibc-rs/pull/1053/commits/9333f775d3670c1f1922eb9b110015db4b7d5330)

File: ibc-apps/ics721-nft-transfer/Cargo.toml (hunk 3 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics721-nft-transfer/Cargo.toml]: [9333f77](https://github.com/cosmos/ibc-rs/pull/1053/commits/9333f775d3670c1f1922eb9b110015db4b7d5330)

**Pull request merged**